pkgs/development/go-modules/generic/default.nix

```diff
@@ -20,6 +20,8 @@
 , vendorSha256 ? null
 # Whether to delete the vendor folder supplied with the source.
 , deleteVendor ? false
+# Lots of packages don't have a good go.mod file, this regenerates it and uses it for downloading & building.
+, regenGoMod ? false
```

Comment:
`regenGoMod` needs to be documented in `doc/languages-frameworks/go.xml`.

Reply:
Is there any harm in running this by default?

Reply:
It might obfuscate the fact, that the `go.mod` should be reported to upstream. On the other hand this put‘s an additional burden on to nixpkgs packagers. So I opt for making it the default.

What I don‘t know is if that might trigger vendor hashes to change on modules that use the module. Someone with more insight into go could perhaps tell, if it is possible to have two different valid `go.mod` files.

Reply:
If this is a potential source for non-reproducibility we should instead of regenerating it just log a diff file that users should include as the patch. Otherwise we are just running danger of breaking hash sums.

Reply:
> Someone with more insight into go could perhaps tell, if it is possible to have two different valid go.mod files.

I think, in general, as long as the contents of vendor contain what the code needs to build, and what's marked as explicit in vendor/modulex.txt matches what's marked as explicit in go.mod - I believe there are an infinite number of potential valid go.mod files for a given set of go code (since you can always add a go module).

>What I don‘t know is if that might trigger vendor hashes to change on modules that use the module

I think there are two interpretations of this question - would we need to regenerate all the hashes? (yes). Would this cause some sort of cascading effect on go code that imports other go code? (no - we don't have go code import other nix packages).

> Is there any harm in running this by default?

It sounds safe, but there is actually a massive problem. Not all go code can run go mod tidy. I think given that only 5 packages need this fix and the rest build either way, I'd prefer to not spend a lot of time updating all the hashes and flipping the go around.

> If this is a potential source for non-reproducibility we should instead of regenerating it just log a diff file that users should include as the patch.

This is not a source of non reproducibility - go mod tidy should always produce the same output and is deterministic based on the source code.

> doc/languages-frameworks/go.xml

I really wish nix used autogenerated documentation from source (or maybe markdown). Either way, updated.

Reply:
I don't understand how it can be reproducible when it only read the source code. What happens if a referenced package is updated after we added the checksum? Would it not than jump to a newer version and therefore break the old checksum?

Reply:
Yeah - if go mod tidy is adding packages, this will cause regeneration to break. I think patches is probably the better solution since it's sticky. Hopefully the conformance of the go ecosystem will improve and we'll stop seeing horribly broken go.mod files.